probe-rs/src/architecture/riscv/communication_interface.rs

```diff
@@ -0,0 +1,848 @@
+        //let b = 9; // base register -> s0
+        //let w = 0; // width
+        //let d = 9; // dest register -> s0
+        //let l = 0b11;
+
+        //let lw_command = bitpack!("oooooooooooobbbbb_www_ddddd_lllllll");
+        let mut sw_command: u32 = 0b0000000_01001_01000_000_00000_0100011;
+
+        // sw command -> sb s1, 0(s0)
+
+        // verify the width is supported
+        // 0 ==  8 bit
+        // 1 == 16 bit
+        // 2 == 32 bit
+
+        assert!(width < 3, "Width larger than 3 not supported yet");
+
+        sw_command |= (width as u32) << 12;
+
+        //if width == 2 {
+        //    sw_command = 0xc004;
+        //}
+
+        let ebreak_cmd = 0b000000000001_00000_000_00000_1110011;
+
+        self.write_dm_register(Progbuf0(sw_command))?;
+        self.write_dm_register(Progbuf1(ebreak_cmd))?;
+
+        // write value into s0
+        self.abstract_cmd_register_write(0x1008, address)?;
+
+        // write address into data 0
+        self.write_dm_register(Data0(data))?;
+
+        // Write s0, then execute program buffer
+        let mut command = AccessRegisterCommand(0);
+        command.set_cmd_type(0);
+        command.set_transfer(true);
+        command.set_write(true);
+
+        // registers are 32 bit, so we have size 2 here
+        command.set_aarsize(2);
+        command.set_postexec(true);
+
+        // register s0, ie. 0x1008
+        command.set_regno(0x1009);
+
+        self.write_dm_register(command)?;
+
+        let status: Abstractcs = self.read_dm_register()?;
+
+        if status.cmderr() != 0 {
+            return Err(RiscvError::AbstractCommand(
+                AbstractCommandErrorKind::parse(status.cmderr() as u8),
+            ));
+        }
+
+        // Restore register s0 and s1
+
+        self.abstract_cmd_register_write(0x1008, s0)?;
+        self.abstract_cmd_register_write(0x1009, s1)?;
+
+        Ok(())
+    }
+
+    pub(crate) fn execute_abstract_command(&mut self, command: u32) -> Result<(), RiscvError> {
+        // ensure that preconditions are fullfileld
+        // haltreq      = 0
+        // resumereq    = 0
+        // ackhavereset = 0
+
+        let mut dmcontrol = Dmcontrol(0);
+        dmcontrol.set_haltreq(false);
+        dmcontrol.set_resumereq(false);
+        dmcontrol.set_ackhavereset(true);
+        dmcontrol.set_dmactive(true);
+        self.write_dm_register(dmcontrol)?;
+
+        // read abstractcs to see its state
+        let abstractcs_prev: Abstractcs = self.read_dm_register()?;
+
+        log::debug!("abstractcs: {:?}", abstractcs_prev);
+
+        if abstractcs_prev.cmderr() != 0 {
+            //clear previous command error
+            let mut abstractcs_clear = Abstractcs(0);
+            abstractcs_clear.set_cmderr(0x7);
+
+            self.write_dm_register(abstractcs_clear)?;
+        }
+
+        self.write_dm_register(Command(command))?;
+
+        // poll busy flag in abstractcs
+
+        let repeat_count = 10;
```

Comment:
Consider using proper time based timeout, or even add backoff handling.

10 retries might be not enough (I think)

Reply:
Adapted to the same time-based timeout as  in `read_dm_register`.